contracts/V1/StateChannelDiamondProxy/StateSnapshotFacet.sol

```diff
@@ -0,0 +1,230 @@
+            channelId,
+            disputeProof
+        );
+
+        _updateStateSnapshot(
+            channelId,
+            milestoneProofs,
+            milestoneSnapshots,
+            exitChannelBlocks,
+            genesisStateSnapshot
+        );
+    }
+
+    function updateStateSnapshotWithoutDispute(
+        bytes32 channelId,
+        ForkMilestoneProof[] memory milestoneProofs,
+        StateSnapshot[] memory milestoneSnapshots,
+        ExitChannelBlock[] memory exitChannelBlocks
+    ) external onlySelf {
+        require(
+            getDisputeLength(channelId) == getSnapshotForkCnt(channelId),
+            "Dispute proof is required"
+        );
+        StateSnapshot memory genesisStateSnapshot = stateSnapshots[channelId];
+
+        _updateStateSnapshot(
+            channelId,
+            milestoneProofs,
+            milestoneSnapshots,
+            exitChannelBlocks,
+            genesisStateSnapshot
+        );
+    }
+
+    function _resolveGenesisSnapshot(
+        bytes32 channelId,
+        DisputeProof memory disputeProof
+    ) internal view returns (StateSnapshot memory) {
+        if (getDisputeLength(channelId) != getSnapshotForkCnt(channelId)) {
+            // dispute proof is required
+            Dispute memory dispute = disputeProof.dispute;
+            uint disputeTimestamp = disputeProof.timestamp;
+
+            require(
+                _isDisputeProofProvided(disputeTimestamp),
+                "Dispute proof is required"
+            );
+            require(
+                _isDisputeCommitmentValid(dispute, disputeTimestamp, channelId),
+                "Dispute proof is not valid"
+            );
+            require(
+                _isDisputeFinalized(disputeTimestamp),
+                "Dispute is not finalized"
+            );
+            require(
+                _isStateSnapshotValid(
+                    disputeProof.outputStateSnapshot,
+                    dispute
+                ),
+                "State snapshot is not valid"
+            );
+
+            return disputeProof.outputStateSnapshot;
+        } else {
+            return stateSnapshots[channelId];
+        }
+    }
+
+    function _updateStateSnapshot(
+        bytes32 channelId,
+        ForkMilestoneProof[] memory milestoneProofs,
+        StateSnapshot[] memory milestoneSnapshots,
+        ExitChannelBlock[] memory exitChannelBlocks,
+        StateSnapshot memory genesisStateSnapshot
+    ) internal {
+        // verify state proof within the fork
+        bool isStateValid = _verifyForkProof(
+            milestoneProofs,
+            milestoneSnapshots,
+            genesisStateSnapshot
+        );
+        require(isStateValid, "Invalid State Proof");
+
+        StateSnapshot memory lastProovenSnapshot = milestoneSnapshots[
+            milestoneSnapshots.length - 1
+        ];
+
+        _validateExitChannelBlocks(
+            exitChannelBlocks,
+            genesisStateSnapshot,
```

Comment:
for exitChannels 'genesisStateSnapshot' should always be onChainStateSnapshot since we're processing withdrawals between the last onChainSnapshot and the current one that we're updating